✨ The first official version

burgholzer opened the pull request:
✨ Starting the new year with a bang 💥
 
This PR contains the work of the last couple of weeks by @ystade and myself on refining QDMI based on the discussions we had at the end of last year. 
Given the discussions in #109, #117 and #125, it was clear that breaking changes were inevitable.
And this PR definitely breaks things. **However**, it also fixes and improves literally all aspects of QDMI. In particular:

- The interfaces have been consolidated and restructured. There are now two clear interfaces: the "device interface" (contained in the `qdmi/device.h` header) and the "client interface" (contained in the `qdmi/client.h` header). These two interfaces connect three entities: devices, clients, and a driver. This distinction and separation of concerns has been made more explicit in the interface itself as well as in its documentation.
- In addition to the regular `QDMI_Session` concept for establishing a connection between a client and a driver, a new `QDMI_Device_Session` concept was introduced that accomplishes a similar connection between the driver and a device. This resolves #109 and #117, based on the discussion in these issues, as it allows the driver to establish separate connections to a device for each client. The respective sessions may be used to authenticate with the device or to potentially switch the API endpoint used in the device implementation without having to recompile the device implementation.
- Resolves #125 by streamlining authentication and session handling. Based on the discussions over there, the new expected workflow with sessions is to first create them, then set certain parameters, and, finally, to initialize them. A session may only be used once it is successfully initialized. Any kind of authentication is handled in the new initialization routines.
- Resolves #128 and supersedes #129 by adding a unique identifier to QDMI sites. Significant effort has been put into clarifying the respective semantics around site IDs and qubits used in programs submitted to the devices.
- Many functions have been renamed to follow a consistent naming scheme throughout the whole interface. Functions are now generally prefixed by the type they operate on (for example, `QDMI_device`; typically the type of the first parameter), followed by the action to be performed (for example, `query`), and suffixed by the particular thing acted on (for example, `site_property`). This style is now consistently used throughout the whole interface.
- The `QDMI_device_get_sites` and `QDMI_device_get_operations` routines have been refactored to become device properties that can be queried just as every other property. This helps to keep the interface more compact and the semantics more uniform.
- Based on discussion around pulse-level control, the property query function for operations has received additional parameters for querying properties of parametrized operations depending on their parameter values. This allows devices to expose almost arbitrary operations and their properties. Effectively, this should allow realizing pulse-level QDMI devices.
- Resolves #108 by adding the respective `calibration` program format and `needs calibration` device property as discussed in that issue.
- Resolves #123 by adjusting the figure as discussed in that issue.
- Resolves #131 by adding extensive API documentation including examples.
- Resolves #106 by significantly expanding on the client interface documentation.
- The API documentation has been meticulously cleaned up and extended wherever possible.
- All example devices, the example driver, as well as the example fomac and the example tool have been updated based on the above changes.
- All tests have been updated based on the above changes.
- ... (I probably forgot a couple of things)

Overall, I believe the list above illustrates the significant progress made since last year's discussions. In particular, this PR closes all open issues requiring breaking changes that we are aware of. Based on that and the level of attention paid to every little detail in this PR, I am confident that, once merged, this will be the first official stable version of QDMI.

> [!NOTE]
> This is a huge pull request. 
> However, given the sheer scope of the changes in this PR, it is not easily possible to split it into smaller chunks. Splitting it up in retrospect would cause significant churn and lead to further delays in the official first release. 
> During development, we took extra care to make small-grained commits so that the commits themself allow to follow the developments and the individual changes.
> As such, the idea would be to _not_ squash the commit history of this PR before merging, but to perform a regular merge that preserves the individual commits.
> Any potential reviewer is advised to look through the git diff in the `examples` folder as this will nicely illustrate the changes on the device and on the client side.

One of the only questions left is the version number we want to use. The two options I see are `v1.0.0` or `v2.0.0`.
Given how we have not made a public 1.0 release yet, it would be justified to call the release `v1`. However, given how people have started to implement QDMI as it is currently on the `develop` branch and how significant the changes in this PR are to the interface, one might rather argue that we should tag the current version on the `develop` branch as `v1` and the version after merging this PR as `v2`. In my personal opinion, that would be the cleaner solution, but I am open to the opinion of others here.
In any case, the version currently on the develop branch will no longer be supported once this PR is merged.

burgholzer committed:
♻️ experimental interface consolidation

Signed-off-by: burgholzer <burgholzer@me.com>

ystade committed:
♻️ Refactor interface

ystade committed:
🐛 Update devivce symbol list

ystade committed:
♻️ Bikeshedding on names in interface

ystade committed:
🐛 Update symbol list

ystade committed:
♻️ Update C device

ystade committed:
♻️ Update defs test for devices

ystade committed:
🎨 Update interface, add session_init

ystade committed:
♻️ Update C device

ystade committed:
🎨 Use switch cases where clearer

ystade committed:
♻️ Update c++ device to current interface

ystade committed:
♻️ Update fomac to new interface

ystade committed:
🎨 Update template to new interface structure

ystade committed:
♻️ Update tests

ystade committed:
♻️ Update tests

ystade committed:
🚧 Wip on driver

ystade committed:
🎨 Store libraries separate from devices

ystade committed:
🐛 Add missing parentheses in C++ device

ystade committed:
🎨 Update dirver with encapsulated types

ystade committed:
🐛 Update defs test

ystade committed:
🐛 Fix some compiler errors

ystade committed:
🐛 Fix a couple of test related bug, add bad state error value

ystade committed:
🐛 Fix some more bugs

ystade committed:
🚸 Use Bad State Error

ystade committed:
🐛 Initialize Device Session in Driver

ystade committed:
🐛 Fix bug in driver

ystade committed:
🎨 Handle readonly mode better in tests

ystade committed:
🐛 Fix bug in job creation

ystade committed:
🐛 Fix bug in setting job param

ystade committed:
🐛 Add mising return/break in switch-case in both devices

ystade committed:
🐛 Fix session set param test

ystade committed:
🐛 Fix free job in driver

ystade committed:
🎨 Fix compiler warning

ystade committed:
🎨 Produce clearer error messages in tests

ystade committed:
🐛 Fix tests

burgholzer committed:
🚧♻️ work in progress refactor based on discussions

Signed-off-by: burgholzer <burgholzer@me.com>

ystade committed:
🐛 Fix C device

ystade committed:
🎨 Remove Op Name prop

ystade committed:
🐛 Fix C++ device

ystade committed:
🎨 Correct error message

ystade committed:
🐛 Fix Test after removing op name prop

ystade committed:
🐛 Fix tests

ystade committed:
🐛 Fix tests

ystade committed:
Revert "🐛 Fix tests"

This reverts commit fe6329fd9a9f9e37c9e604a2b216fea5caad7565.

ystade committed:
Revert "🐛 Fix Test after removing op name prop"

This reverts commit 407394d93c9c28e4a3c63be68ab87cb976ee6f4c.

ystade committed:
Revert "🎨 Correct error message"

This reverts commit 93c34e5de581b83d06fa3638c99d5f07bfed7c27.

ystade committed:
Revert "🐛 Fix C++ device"

This reverts commit 84f880eb3ddc4b2438ea46260fc9ae77f1e0bf8e.

ystade committed:
Revert "🎨 Remove Op Name prop"

This reverts commit 8408743cff81512e7dbf41700af110b80c4bc8b9.

ystade committed:
Revert "🐛 Fix C device"

This reverts commit 92a41c4e13f04d1cfd8423c8438cd8ea09192646.

ystade committed:
Revert "🚧♻️ work in progress refactor based on discussions"

This reverts commit 57e1e84054446886d9aa94e6b29c6cbf29cf3419.

ystade committed:
🎨 Re-introduce opaque pointers for sites and ops in interface

ystade committed:
🐛 Fix interface

ystade committed:
🐛 Fix interface

ystade committed:
🎨 Update and fix c device

ystade committed:
🎨 Update driver

ystade committed:
🎨 Update FoMac

ystade committed:
💚 Fix clang tidy warning

ystade committed:
🎨 Update C++ device

ystade committed:
🐛 Update template and fix several bugs

ystade committed:
🎨 Disable C++ tests for now

ystade committed:
♻️ Clean-up interface

ystade committed:
🐛 Fix Name shifting

ystade committed:
🐛 Fix bugs

ystade committed:
🏗️ Do not define structs in headers

burgholzer committed:
📝 write proper docstring for `types.h`

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 improve device session docs

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🩹 fix duplicate enum value

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✏️ plural

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 clarify who sets device job parameters

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 small tweaks to device property docstrings

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✨ introduce site ID property

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 document expected program qubit to site mapping

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🩹 device query functions all need a session parameter

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✨ implement the site ID concept

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
♻️ replace `get_sites` and `get_operations` with device properties

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🩹 make docs compile again

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🎨 consistent naming scheme throughout QDMI

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 improved documentation for parameter setting routines

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🩹 interface conforming setters

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🎨 code style

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
♻️ make program format and program job parameters and fix naming inconsistency in `QDMI_JOB_PARAMETER_SHOTS_NUM`

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🎨 cleaning up the driver implementation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🎨 order the methods in the interface according to the expected call sequence

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
♻️ replace `get_devices` method with `query_session_property` to streamline interface

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
♻️ adjust permission handling in driver session initialization

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 expand on session init docs

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 further refinement in docsstrings

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 run grammar checker on constants.h

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🎨 adjust name of "topics" in docs

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🚨 more warnings in doxygen build

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝🔧 do not sort member docs

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 rework client interface documentation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 refine constants documentation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 rework device interface documentation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🩹 querying with `value` and `size_ret` being NULL is valid

It can be used to check whether a given property is supported at all.

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✨ add list of floating point parameters to operation property query function

This is the result of the pulse-level control discussions, where the need has come up to report properties for operations that depend on the values of some parameters.

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 further small docstring refinements

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✏️ fix american english "cancelled" -> "canceled"

Signed-off-by: burgholzer <burgholzer@me.com>

ystade committed:
📝 Add three todos

ystade committed:
🎨 Fix links in rationale, fix typo, more expressiv tag

ystade committed:
✏️ Fix typo

ystade committed:
📝 Update schematic

ystade committed:
🚸 Improve change detection in CMakeLists.txt

ystade committed:
🎨 Update rationale

ystade committed:
🐛 Fix example page

ystade committed:
🎨 To be reverted see commit msg

The style of the listing of the nested groups on the interface page is suboptimal since it is made for displaying code. This commit contains some css code that would enable better styling IF one would be able to add HTML tag with the class qdmi_interface as a sibling into those HTML elements.

ystade committed:
Revert "🎨 To be reverted see commit msg"

This reverts commit 2001daaa60417a5bdc9211e59cdb46f2e70ff23f.

burgholzer committed:
📝 add to site and operation documentation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 adjustments to constants documentation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝🔧 show enum values in documentation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 adjust file briefs

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🔧 fine tune documentation page structure

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 small refinements to the device interface documentation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 small refinements to the client interface documentation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 small refinements to FAQ

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 small refinements to support text

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 small refinements to development guide

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 small refinements to template description

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 small refinements to examples

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝♻️ rewrite rationale

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🍱 switch job and query interface in schematic

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🍱 remove backloops in QDMI overview sketch

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✨ enable calibration as a type of program

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🩹 ensure the html directory exists for doxygen

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🚨 use C++ memory management in C++ device implementation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🩹 properly initialize and free the `job->program` in the C device implementation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🩹 prevent null dereference in C device implementation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🚨 fix most of the CodeQL warnings

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🚨 fix remaining CodeQL warnings

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✅ adjust tests to account for new `get_parameters_num` fomac method

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🔥 remove unreachable checks

No jobs can be created in READONLY mode, so a client cannot gain access to a `QDMI_Job` handle in that case.

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🐛 fix for undefined behavior in example driver shutdown

it is undefined behavior to modify a container while it is iterated over in C++.

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✅ add a test for `QDMI_session_query_session_property` to test its semantics

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✅ add a test for `QDMI_session_init` to test its semantics

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
♻️ simplify `throw_if_error` implementation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🎨 fix leftover naming inconsistency in T1 and T2 time site properties

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✅ also test submission of calibration jobs

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✅ add FoMaC functions for T1 and T2 times and corresponding tests.

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
👷📝 add PR preview functionality for docs

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🎨 more stringent invalid argument checks in example devices

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✅ adjust tests to check for unsupported custom properties and cover more corner cases

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🎨 refine validity checks in example driver

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
✅ add further tests for corner cases

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer commented:
Alright, it took a while, but the entire CI is as green as it gets ✅ 
I also added a nice little feature to build documentation previews for PRs so that people can have a look at the documentation before it is merged, without having to build it locally.

The preview for this PR is available at https://munich-quantum-software-stack.github.io/QDMI/pr-preview/pr-132/
This is probably the best way of previewing the changes from this PR.

I will keep this PR open for comments until sometime next week. If nothing big pops up until then, I will go ahead and merge the PR and proceed with the release.

I am explitly tagging a couple of people here that were involved in the related discussions.
@ystade @echavarria-lrz @mnfarooqi @AdVetter @denialhaag

robertwille commented:
Wow, this is truly impressive!👍 Huge thanks @burgholzer and @ystade for all your efforts on this essentiell component!🙏

ystade commented:
Regarding the versioning: I would actually be in favor of v1.1.0. I agree that within the interface we made a lot of changes. However, from the perspectives of users of QDMI, e.g., device implementers, I think the changes are still moderate. In the example device we only moved some pieces of code and renamed a couple of functions but not more. Additionally, I am hesitant to leave version v1.0.0 behind because it is not working (see all related issues). In my opinion, we should only start with a new version when we have fixed the old one such that it is working, which is not true for v1 right now.

ystade reviewed (approved):
This PR comes with plenty of changes but I think all of them for the good. Everything looks good to me.

burgholzer committed:
🔖 marking as version 1.1.0

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer commented:
> Regarding the versioning: I would actually be in favor of v1.1.0. I agree that within the interface we made a lot of changes. However, from the perspectives of users of QDMI, e.g., device implementers, I think the changes are still moderate. In the example device we only moved some pieces of code and renamed a couple of functions but not more. Additionally, I am hesitant to leave version v1.0.0 behind because it is not working (see all related issues). In my opinion, we should only start with a new version when we have fixed the old one such that it is working, which is not true for v1 right now.

Totally agree with the reasoning. Adjusted the version number in https://github.com/Munich-Quantum-Software-Stack/QDMI/pull/132/commits/d64b4eeab54c4e786aee0ce126eaacadfa247656.
I will tag https://github.com/Munich-Quantum-Software-Stack/QDMI/commit/fa142a656bbed226ba4e38051d00931f9656c2e9 as `v1.0.0` since it was originally intended as such and no _significant_ (breaking) changes have been committed to the `develop` branch since then. Most likely, I will also create an official GitHub release for that so that it does not look as odd when the releases start with `v1.1.0`.
Then, I'll merge this PR and tag the respective commit as `v1.1.0` with a corresponding release.

denialhaag commented on docs/guide.md:
```diff
@@ -142,22 +141,23 @@ and `<PATH_TO_INCLUDE_DIRECTORY>` is the path to the `include` directory of the
-For the information to be displayed correctly in the documentation, it is essential that the
-comments follow the format required by Doxygen. Below you find some tags, that are commonly used
-within the documentation of a function:
+For the information to be displayed correctly in the documentation, it the comments need to follow
```
```suggestion
For the information to be displayed correctly in the documentation, the comments need to follow
```

denialhaag commented on docs/templates.md:
```diff
@@ -31,32 +31,32 @@ device. The following sections describe how to set up and use the template.
+file in the root directory of the template.
 
-For stability, we recommend to pin the version of QDMI that you are using for your implementation.
+For stability, we recommend pining the version of QDMI that you are using for your implementation.
```
```suggestion
For stability, we recommend pinning the version of QDMI that you are using for your implementation.
```

denialhaag commented on docs/templates.md:
```diff
@@ -65,10 +65,10 @@ variable in `cmake/ExternalDependecies.cmake` as follows:
 +   set(QDMI_REV "v1.0.0"
```
Should this reflect the most recent version, that is, v1.1.0?

burgholzer replied:
Yeah. Thanks for caching this!
```suggestion
+   set(QDMI_REV "v1.1.0"
```

denialhaag commented on examples/device/cxx/device.cpp:
```diff
@@ -655,19 +651,196 @@ int CXX_QDMI_control_get_data_dev(CXX_QDMI_Job job,
+                      prop, size, value, size_ret)
+  ADD_STRING_PROPERTY(QDMI_DEVICE_PROPERTY_VERSION, "0.1.0", prop, size, value,
+                      size_ret)
+  ADD_STRING_PROPERTY(QDMI_DEVICE_PROPERTY_LIBRARYVERSION, "1.0.0", prop, size,
```
```suggestion
  ADD_STRING_PROPERTY(QDMI_DEVICE_PROPERTY_LIBRARYVERSION, "1.1.0", prop, size,
```

denialhaag reviewed (commented):
Thanks for the quick implementation! The changes address our problems raised in #109. We look forward to updating our device when this PR is merged.

While looking at the docs and `device.cpp`, I found some minor typos. Otherwise, the changes look good to me.

burgholzer commented on examples/device/c/device.c:
```diff
@@ -408,243 +385,432 @@ int C_QDMI_control_check_dev(C_QDMI_Job job, QDMI_Job_Status *status) {
+                      size, value, size_ret)
+  ADD_STRING_PROPERTY(QDMI_DEVICE_PROPERTY_VERSION, "0.1.0", prop, size, value,
+                      size_ret)
+  ADD_STRING_PROPERTY(QDMI_DEVICE_PROPERTY_LIBRARYVERSION, "1.0.0", prop, size,
```
```suggestion
  ADD_STRING_PROPERTY(QDMI_DEVICE_PROPERTY_LIBRARYVERSION, "1.1.0", prop, size,
```

burgholzer committed:
✏️ fix a couple of typos and missed version updates

Co-authored-by: Daniel Haag <121057143+denialhaag@users.noreply.github.com>
Signed-off-by: Lukas Burgholzer <burgholzer@me.com>

burgholzer commented:
> Thanks for the quick implementation! The changes address our problems raised in #109. We look forward to updating our device when this PR is merged.
> 
> While looking at the docs and `device.cpp`, I found some minor typos. Otherwise, the changes look good to me.

Many thanks for catching these! Fixed in https://github.com/Munich-Quantum-Software-Stack/QDMI/pull/132/commits/188dc185c6b58cc35e513248e156c42dfda3960e ✅

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

mnfarooqi commented:
@burgholzer @ystade 
Nice work with addressing all the outstanding issues and organising the interface in one go.

Although I have an idea why the following might be there, it would be better if you could clarify the following: 
1) What's the reason for putting three enums (constants) in client.h when all the others are in `constants.h`? 
2) Why are there two job parameter enums `QDMI_JOB_PARAMETER_T` in `client.h` and `QDMI_DEVICE_JOB_PARAMETER_T` in `constants.h`?

burgholzer committed:
📝 add job parameter explanation to rationale

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
⬆️🪝 update pre-commit hooks

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
📝 add rationale for enumeration placement

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer commented:
> @burgholzer @ystade Nice work with addressing all the outstanding issues and organising the interface in one go.
> 
> Although I have an idea why the following might be there, it would be better if you could clarify the following:
 
Thanks for raising these questions. I added two commits that hopefully address these to the rationale section of the documentation. You can preview them under the links below.

> 1. What's the reason for putting three enums (constants) in client.h when all the others are in `constants.h`?

https://munich-quantum-software-stack.github.io/QDMI/pr-preview/pr-132/md_docs_2rationale.html#rationale-job-structs

> 2. Why are there two job parameter enums `QDMI_JOB_PARAMETER_T` in `client.h` and `QDMI_DEVICE_JOB_PARAMETER_T` in `constants.h`?

https://munich-quantum-software-stack.github.io/QDMI/pr-preview/pr-132/md_docs_2rationale.html#rationale-enum-definitions

Feel free to suggest edits for any of the entries ☺️

echavarria-mqv commented:
Awesome work guys. I have no major concerns on my end. The PR looks good to go, please feel free to merge it.

burgholzer commented:
Alright. Thank you, everyone, for your feedback! Looks like this is good to go. 🚀 
I will kick off the release process tomorrow and, finally, close this first chapter in the history of QDMI 📖

burgholzer committed:
✏️ slightly reword readme sentence

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
⬆️🪝 update pre-commit hooks

Signed-off-by: burgholzer <burgholzer@me.com>
